route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,346 @@
+                    0,
+                    feature_list_.size() - 1);
+                std::uniform_int_distribution<int> delay_distribution(
+                    500,
+                    1500);
 
-    Notification<Status> done;
-    client_->Call<RouteGuide, Stream<Point>, RouteSummary>(
-        "RecordRoute",
-        [&](auto&& call, bool ok) {
-          call->OnFinished([&](auto*, const Status& status) {
-            done.Notify(status);
-          });
-          if (!ok) {
-            call->context()->TryCancel();
-            call->Finish();
-          } else {
-            for (int i = 0; i < kPoints; i++) {
-              const Feature& f = feature_list_[feature_distribution(generator)];
-              std::cout << "Visiting point "
-                        << f.location().latitude()/kCoordFactor_ << ", "
-                        << f.location().longitude()/kCoordFactor_ << std::endl;
-              if (call->Write(f.location()) != ClientCallStatus::Ok) {
-                // Broken stream.
-                break;
-              }
-              std::this_thread::sleep_for(std::chrono::milliseconds(
-                  delay_distribution(generator)));
-            }
-            call->WritesDone();
-            call->OnRead([&](auto* call, auto&& response) {
-              stats.Swap(response.get());
-              call->Finish();
-            });
-          }
-        });
-    Status status = done.Wait();
-    if (status.ok()) {
-      std::cout << "Finished trip with " << stats.point_count() << " points\n"
-                << "Passed " << stats.feature_count() << " features\n"
-                << "Travelled " << stats.distance() << " meters\n"
-                << "It took " << stats.elapsed_time() << " seconds"
-                << std::endl;
-    } else {
-      std::cout << "RecordRoute rpc failed." << std::endl;
-    }
+                for (int i = 0; i < data.num_points; i++) {
+                  const Feature& f =
+                      feature_list_[feature_distribution(generator)];
+                  std::cout << "Visiting point "
+                            << f.location().latitude() / kCoordFactor_
+                            << ", "
+                            << f.location().longitude() / kCoordFactor_
+                            << std::endl;
+                  call.Write(f.location());
+                  std::this_thread::sleep_for(std::chrono::milliseconds(
+                      delay_distribution(generator)));
+                }
+                // When writes are done we should call WritesDone in order
+                // to finish our call.
+                call.WritesDone();
+              })
+
+              // The .body() callback is invoked every time the server
+              // sends a response.
+              .body([](auto&& data, auto& call, auto&& response) {
+                // We should check if the response is valid from the server.
+                if (!response) {
```

Comment:
As with the comment below about type-safe checks:
```suggestion
                if (response == nullptr) {
```